security/src/main/java/pm/gnosis/svalinn/security/impls/AesEncryptionManager.kt

```diff
@@ -208,38 +195,33 @@ class AesEncryptionManager(
                 }
             }
 
-    override fun observeFingerprintForUnlock(): Observable<FingerprintUnlockResult> =
-        Observable
-            .fromCallable {
-                val cryptedData = preferencesManager.prefs.getString(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY, null) ?: throw FingerprintUnlockError()
-                CryptoData.fromString(cryptedData)
-            }
-            .flatMap { cryptedData -> fingerprintHelper.authenticate(cryptedData.iv).map { cryptedData to it } }
-            .subscribeOn(Schedulers.io())
-            .map { (cryptedData, authResult) ->
-                when (authResult) {
-                    is AuthenticationResultSuccess -> {
-                        synchronized(keyLock) {
-                            key = authResult.cipher.doFinal(cryptedData.data)
-                        }
-                        if (key != null) FingerprintUnlockSuccessful() else throw FingerprintUnlockError()
-                    }
-                    is AuthenticationFailed -> FingerprintUnlockFailed()
-                    is AuthenticationHelp -> FingerprintUnlockHelp(authResult.helpString)
+    override suspend fun unlockWithFingerprint(): FingerprintUnlockResult {
+        val cryptedData = CryptoData.fromString(
+            preferencesManager.prefs.getString(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY, null) ?: throw FingerprintUnlockError()
+        )
+        val authResult = fingerprintHelper.authenticate(cryptedData.iv)
+        return when (authResult) {
+            is AuthenticationResultSuccess -> {
+                synchronized(keyLock) {
+                    key = authResult.cipher.doFinal(cryptedData.data)
                 }
+                if (key != null) FingerprintUnlockSuccessful() else throw FingerprintUnlockError()
             }
+            is AuthenticationFailed -> FingerprintUnlockFailed()
+            is AuthenticationHelp -> FingerprintUnlockHelp(authResult.helpString)
+        }
+    }
 
-    override fun isFingerPrintSet(): Single<Boolean> =
-        fingerprintHelper.isKeySet()
-            .map { it && preferencesManager.prefs.getString(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY, null) != null }
-            .onErrorReturn { false }
+    override fun isFingerPrintSet(): Boolean =
+        nullOnThrow { fingerprintHelper.isKeySet() } == true &&
+                preferencesManager.prefs.getString(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY, null) != null
 
-    override fun clearFingerprintData(): Completable =
-        Completable.fromCallable {
-            preferencesManager.prefs.edit {
-                remove(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY)
-            }
-        }.andThen(fingerprintHelper.removeKey())
+    override fun clearFingerprintData() {
+        preferencesManager.prefs.edit {
+            remove(PREF_KEY_FINGERPRINT_ENCRYPTED_APP_KEY)
+        }
+        fingerprintHelper.removeKey()
```

Comment:
My understanding is that `Completable.andThen` only executes on success. If for some reason `remove()` fails ...we still throw and everything behaves the same, right?

Reply:
We have a test for this: expected behaviour is that is throws